### PR TITLE
Implement tabs and spaces consistency check

### PR DESCRIPTION
**Pull request by rediska0123**
@a-litvinov, can you please review this?

### CODE REVIEW

File: src/linter.cc [hunk 1]
```diff
@@ -240,6 +240,59 @@ absl::Status CheckAliasKeyword(absl::string_view sql) {
...
+    std::vector<CodePosition> tabs, spaces;
+    for (int i=0; i<static_cast<int>(sql.size()); ++i) {
+        if ( sql[i] == delimeter ) {
+            ++line_number;
+            line_size = 0;
+            is_indent = true;
+            continue;
+        } else if ( sql[i] == space ) {
+            if ( is_indent ) {
+                spaces.push_back({ line_number, line_size });
```
Comment (a-litvinov) [inline src/linter.cc:259]: Why do you save all spaces and tabs? I think just a first occurrence of each is enough, no?

File: src/linter.cc [hunk 2]
```diff
@@ -240,6 +240,59 @@ absl::Status CheckAliasKeyword(absl::string_view sql) {
...
+    bool is_indent = true;
+    std::vector<CodePosition> tabs, spaces;
+    for (int i=0; i<static_cast<int>(sql.size()); ++i) {
+        if ( sql[i] == delimeter ) {
+            ++line_number;
+            line_size = 0;
+            is_indent = true;
+            continue;
+        } else if ( sql[i] == space ) {
+            if ( is_indent ) {
```
[the remark this answers is not in the document]
Comment (rediska0123) [reply src/linter.cc:258]: We discussed this with Andrey and finally thought that we should make an another check for the tab characters that are not at the start

File: src/linter_test.cc [hunk 3]
```diff
@@ -91,5 +91,19 @@ TEST(LinterTest, AliasKeywordCheck) {
     EXPECT_TRUE(CheckAliasKeyword("SELECT 1 AS one").ok());
 }
 
+TEST(LinterTest, TabCharactersUniformCheck) {
```
Comment (nastarann) [inline src/linter_test.cc:94]: Can you also add test cases for negative cases which also check the error position reported? E.g.

EXPECT_FALSE(CheckTabCharactersUniform("  SELECT 5;\n  \tSELECT 6;").ok());

File: src/linter.cc [hunk 4]
```diff
@@ -240,6 +243,40 @@ absl::Status CheckAliasKeyword(absl::string_view sql) {
     }).ApplyTo(sql);
 }
 
+absl::Status CheckTabCharactersUniform(absl::string_view sql,
+    const char indent, const char delimeter) {
```
Comment (a-litvinov) [inline src/linter.cc:251]: `CheckIndentCharacters(absl::string_view sql, const char allowed_indent, const char line_delimeter)` ?

File: src/linter.cc [hunk 5]
```diff
@@ -240,6 +243,40 @@ absl::Status CheckAliasKeyword(absl::string_view sql) {
...
+    bool is_indent = true;
+    const char kSpace = ' ', kTab = '\t';
+
+    for (int i=0; i<static_cast<int>(sql.size()); ++i) {
+        if ( sql[i] == delimeter ) {
+            is_indent = true;
+        } else if ( sql[i] != indent ) {
```
Comment (a-litvinov) [inline src/linter.cc:258]: please move is_indent check to the outer if, i.e.:
`esle if (is_indent && sql[i] != allowed_indent) { ...` 

This way the code is easier to read (at least for me :)

File: src/linter.cc [hunk 6]
```diff
@@ -240,6 +243,40 @@ absl::Status CheckAliasKeyword(absl::string_view sql) {
...
+            if ( (sql[i] == kTab || sql[i] == kSpace) && is_indent ) {
+                // Return error on position <i>.
+                ParseLocationPoint lp =
+                    ParseLocationPoint::FromByteOffset(i);
+                ParseLocationTranslator lt(sql);
+                zetasql_base::StatusOr <std::pair <int, int> > status_or_pos =
+                    lt.GetLineAndColumnAfterTabExpansion(lp);
+                ZETASQL_RETURN_IF_ERROR(status_or_pos.status());
+                std::pair <int, int> error_pos = status_or_pos.value();
```
Comment (a-litvinov) [inline src/linter.cc:267]: There is another convenience macro that allows writing this in 1 line:
`ZETASQL_ASSIGN_OR_RETURN(std::pair<int, int> pos, lt.GetLineAndColumnAfterTabExpansion(lp));`

Comment (rediska0123) [reply src/linter.cc]: I couldn't figure out why this doesn't work with `std::pair <int, int> error_pos` but works with `auto error_pos` :(
It says
```
src/linter.cc:264:58: error: expected primary-expression before 'int'
  264 |                 ZETASQL_ASSIGN_OR_RETURN(std::pair <int, int> error_pos,
      |                                                          ^~~
```

**Comment by rediska0123** [comment]
> you should be able to write just `indent` instead of `std::string(1, indent)` here.

It says 
```
external/com_google_absl/absl/strings/str_cat.h:257:3: note: declared here
  257 |   AlphaNum(char c) = delete;  // NOLINT(runtime/explicit)
      |   ^~~~~~~~
```
So I thought maybe characters are not supported

File: src/linter.cc (hunk 6 above)
Comment (a-litvinov) [reply src/linter.cc]: ah, sorry, my bad. This macro doesn't play well with <> characters :( How about:
```
std::pair<int, int> pos;
ZETASQL_ASSIGN_OR_RETURN(pos, lt.GetLineAndColumnAfterTabExpansion(lp));
```

Comment (rediska0123) [reply src/linter.cc]: This works, thanks!
